### PR TITLE
Script to pull unique run logs from ABR robots

### PR DESCRIPTION
**Pull request by rclarke0**
# Overview

This folder pulls run log information from each robot in ABR and saves them locally to your personal computer

# Test Plan

Example Test Plan (HTTP API Change)

- Pulled run logs from all ABR robots with and without robots running
- Concluded that pulling while running slows the robots movements down but does not cause any failed runs to occur
- Pulled multiple times in a row to ensure duplicate runs were not being added and new runs were not being missed

# Changelog

- Added ABR IP addresses
- Added script to pull command, health, and protocol data using http requests with each ip address
- saves runs with unique run ids

# Review requests

N/A

# Risk assessment
- IP Addresses are now public to the internet

This script has more edits to be made:
- if IP addresses on ABR robots change the script will need to be updated
- Eventually the script will be edited so that it no longer saves to a local computer
- Eventually the script will include more data extraction from the robot including: labware offsets per run, instrument serials per run, etc.

### CODE REVIEW

File: hardware-testing/hardware_testing/abr_tools/abr_run_logs.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+def save_runs(runs_to_save: Set[str], ip: str, storage_directory: str) -> None:
+    for a_run in runs_to_save:
+        data = get_run_data(a_run, ip)
+        robot_name = data["robot_name"]
+        data_file_name = data["robot_name"] + "_" + data["run_id"] + ".json"
+        json.dump(data, open(os.path.join(storage_directory, data_file_name), mode="w"))
+    print(
+        f"Saved {len(runs_to_save)} run(s) from robot {robot_name} with IP address {ip}."
+    )
+    pass
```
Comment (sfoster1) [inline hardware-testing/hardware_testing/abr_tools/abr_run_logs.py]: should remove the `pass`, not necessary anymore

File: hardware-testing/hardware_testing/abr_tools/abr_run_logs.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    runs_from_storage = get_run_ids_from_storage(storage_directory)
+    for ip in ABR_IPS:
+        try:
+            runs = get_run_ids_from_robot(ip)
+            runs_to_save = get_unseen_run_ids(runs, runs_from_storage)
+            save_runs(runs_to_save, ip, storage_directory)
+        except Exception:
+            print(f"Failed to read IP address: {ip}.")
+            traceback.print_exc()
+    pass
```
Comment (sfoster1) [inline hardware-testing/hardware_testing/abr_tools/abr_run_logs.py]: should remove the `pass`, not necessary anymore

**Review by sfoster1 — APPROVED** [review]
Looks good to me!

File: hardware-testing/hardware_testing/abr_tools/abr_run_logs.py (hunk 2 above)
Comment (rclarke0) [reply hardware-testing/hardware_testing/abr_tools/abr_run_logs.py]: removed.

File: hardware-testing/hardware_testing/abr_tools/abr_run_logs.py (hunk 1 above)
Comment (rclarke0) [reply hardware-testing/hardware_testing/abr_tools/abr_run_logs.py]: removed.

**Pull request merged**